llvm/lib/Transforms/Utils/FunctionComparator.cpp

```diff
@@ -836,6 +837,54 @@ int FunctionComparator::cmpValues(const Value *L, const Value *R) const {
   return cmpNumbers(LeftSN.first->second, RightSN.first->second);
 }
 
+int FunctionComparator::cmpDistinctMetadata(const Function *L,
+                                            const Function *R) {
+  // Iterate over the basic blocks and instructions in Function L
+  for (const BasicBlock &BBL : *L) {
+    for (const Instruction &InstL : BBL) {
+      for (unsigned i = 0, e = InstL.getNumOperands(); i != e; ++i) {
+        if (MetadataAsValue *MDL =
+                dyn_cast<MetadataAsValue>(InstL.getOperand(i))) {
+          if (MDNode *MDNodeL = dyn_cast<MDNode>(MDL->getMetadata())) {
+            // Now iterate over the basic blocks and instructions in Function R
+            // to compare against each distinct metadata value in Function L
+            for (const BasicBlock &BBR : *R) {
+              for (const Instruction &InstR : BBR) {
```

Comment:
This looks quadratic (compare every operand of every instruction of every basic block in F1 against every operand of ... in F2). Instead, I think you want to iterate in parallel.